### PR TITLE
Loading cache improve eviction use policy v2

### PR DESCRIPTION
**Pull request by vladzcloudius**
This is a v2 of https://github.com/scylladb/scylla/pull/9398

Please, only review - don't merge.
When it's ready for merge I'll squash vX patches and create a clean PR.

Not updating the PR description for the same reason.

### CODE REVIEW

**Comment by vladzcloudius** [comment]
@nyh Please, take a look now.

**Comment by vladzcloudius** [comment]
@nyh ping

**Comment by vladzcloudius** [comment]
@nyh @psarna @tgrabiec ping #2

**Comment by psarna** [comment]
v2 looks good to me, I guess @nyh should approve whether he's satisfied with the fixes he requested

**Comment by vladzcloudius** [comment]
> v2 looks good to me, I guess @nyh should approve whether he's satisfied with the fixes he requested

I think so too! But I can't get him looking at this for 2 weeks by now. :D
Didn't know what to do, hence tagged you, guys.

**Comment by vladzcloudius** [comment]
@nyh ping # 3
@slivne @avikivity Please, advice.

File: cql3/query_processor.cc
Comment (nyh) [inline cql3/query_processor.cc]: As I noted before, I think this user-facing counter needs an easier to understand explanation, not using the phrases "new generation" or "usually a sign of cache polution". Why not say something explicit like "Counts the number of prepared statements that were evicted from the prepared-statement cache before being used even once"? (or twice or whatever the threshold really means, I had a question about this above).

File: cql3/query_processor.cc
Comment (nyh) [inline cql3/query_processor.cc]: Same comment I had above on the phrasing, also here.

File: cql3/prepared_statements_cache.hh
Comment (nyh) [inline cql3/prepared_statements_cache.hh:109]: I don't remember ever asking to rename the "new generation" with "unprivileged". I actually think it's a worse name.

I do prefer the name "section" instead of "partition".

**Review by nyh — APPROVED** [review]
Looks good to me. I did leave a few more comments about comments, naming and strings which I think should be improved - but it's already easier to understand than the previous version. Thanks.

File: cql3/prepared_statements_cache.hh
Comment (nyh) [reply cql3/prepared_statements_cache.hh:115]: @vladzcloudius this is  the point that worries me the most, please address it. I'm worried that I'm not sure if there's something wrong in the comment, in the code or in both (or neither ;-)).

File: cql3/prepared_statements_cache.hh
Comment (vladzcloudius) [reply cql3/prepared_statements_cache.hh:109]: https://github.com/scylladb/scylla/pull/9398#discussion_r718589824

And it was me who suggested it. The "privileged" comes from the fact that entries in this section are going to be evicted only after there are no entries in the "unprivileged" section to evict.

File: cql3/query_processor.cc
Comment (vladzcloudius) [reply cql3/query_processor.cc]: You are commenting about the non-existing code. Please, look at "v2: ..." patches or at the final version of files.

I deliberately left original patches so that you can review only the changes.

File: cql3/query_processor.cc
Comment (vladzcloudius) [reply cql3/query_processor.cc]: https://github.com/scylladb/scylla/pull/9591#discussion_r755332307

File: cql3/prepared_statements_cache.hh
Comment (nyh) [reply cql3/prepared_statements_cache.hh:109]: Ok. Your commit message suggests these are changes which asked for, yet a big chunk of the changes are renaming "new generation" to "unprivileged" which I never asked for, so it was confusing :-)

File: cql3/query_processor.cc
Comment (nyh) [reply cql3/query_processor.cc]: I don't understand, you are saying that you changed the text of this message?
No, renaming "new generation" to "unprivileged" isn't what I'm asking for. A user also won't understand the word "unprivileged". Please read my comment again and check if it really doesn't apply to your newest version (if I doesn't, I don't understand what I'm missing).

File: cql3/prepared_statements_cache.hh
Comment (vladzcloudius) [reply cql3/prepared_statements_cache.hh:115]: @nyh Please, read the comment here (https://github.com/scylladb/scylla/pull/9591/files#diff-938c7e6e78f976ea6c66c4e275a14f8342d56ca1b45a8c14be37dc9695ab0a31R423) and the corresponding code below.

Then entry is going to move to the privileged section only after the amount of "touches" is (strictly) greater than the threshold.

Comment (vladzcloudius) [reply cql3/prepared_statements_cache.hh:115]: There is also a test that checks exactly that: https://github.com/scylladb/scylla/pull/9591/files#diff-a48d906141ec735d4331a48027325efeae331e93066bbb384d9a447783dfb067R397

File: utils/loading_cache.hh
Comment (vladzcloudius) [reply utils/loading_cache.hh]: There is: https://github.com/scylladb/scylla/pull/9591/files#diff-938c7e6e78f976ea6c66c4e275a14f8342d56ca1b45a8c14be37dc9695ab0a31R89

File: cql3/query_processor.cc
Comment (vladzcloudius) [reply cql3/query_processor.cc]: Sorry. NP. I'll change the description.

File: cql3/prepared_statements_cache.hh
Comment (vladzcloudius) [reply cql3/prepared_statements_cache.hh:109]: The series contains not *only* changes that you've requested. In particular there is a bug fix (the last patch) that wasn't part of v1 at all.

File: cql3/prepared_statements_cache.hh
Comment (nyh) [reply cql3/prepared_statements_cache.hh:115]: Vlad, I'm running out of patience for this review. This link you sent to "please read the comment" doesn't work, please answer my comments directly and don't point me to a thousand other places. The test you linked to again has code - but not *text* explaining what the intention is.

The code *looked* to me when I reviewed it that it incremented the counter before checking, so if the counter is 1, and the item is touched it becomes 2 and if the threshold is 2, it is immediately moved to the old generation. That's what I understood, but it's very possible I understood wrong. That's what comments are for... There is also code that checks <threshold and other code that checks >threshold (and surprisingly neither of them has = - it's strict inequalities), so it's confusing. This needs to be explained clearly. It's too confusing.

File: cql3/query_processor.cc
Comment (vladzcloudius) [reply cql3/query_processor.cc]: Pushed a patch that updates the metrics description. Please, review.

File: cql3/prepared_statements_cache.hh
Comment (vladzcloudius) [reply cql3/prepared_statements_cache.hh:115]: Nadav, the code increments the counter *after* it's checked - and this is the source of your confusion.
The link was supposed to get you to the corresponding place in the code - I'm sorry it didn't work.

Please, find the function 'void touch_lru_entry_2_sections(ts_value_lru_entry& lru_entry)' in the `loading_cache.hh`: line 427

This function has a corresponding explanation as well.
Please, let me know if that helps.

File: cql3/prepared_statements_cache.hh
Comment (vladzcloudius) [reply cql3/prepared_statements_cache.hh:109]: But if you referred THIS patch comment then, yeah, it contains fixes addressing your comments and a particular change has been communicated to you in v1 and you didn't object, so... ;)

File: cql3/prepared_statements_cache.hh
Comment (vladzcloudius) [reply cql3/prepared_statements_cache.hh:115]: And the description of the `loading_cache` class has the corresponding description too.
Please, pay a careful attention to boundaries condition in the description that clearly state that the entry moves to a privileged section when the amount of hits is *greater than* the threshold.

File: utils/loading_cache.hh
Comment (vladzcloudius) [reply utils/loading_cache.hh]: This link does work too. :(
See loading_cache.hh: line 89

**Comment by vladzcloudius** [comment]
> 

@nyh Your comments have been addressed (pushed a v3 patch). Please, let me know if the series is good and I can rebase and squash?

File: cql3/query_processor.cc
Comment (nyh) [inline cql3/query_processor.cc:354]: A "red flag" for what? Why?
If I understand correctly, it means the user might be preparing a new statement on each request instead of reusing a previously prepared statement?

**Review by nyh — APPROVED** [review]
I am choosing "approve" because I think it's good enough and I'm tried making the same comments every time. None of my remaining comments are particularly important.

File: cql3/query_processor.cc
Comment (vladzcloudius) [reply cql3/query_processor.cc:354]: Yes, and if he was preparing the same query exactly (e.g. "SELECT * from table where key = ?", then it would not cause a cache eviction but rather a "cache hit" since prepared statements are cached by a query string name.

However if queries are being evicted right after they were used only once means that they are actually used only once, e.g. as described in #8674. 

And this is exactly the opposite from what's the intention of prepared statements and hence it's a "red flag".

**Comment by nyh** [comment]
@vladzcloudius please clean up whatever you want to clean up. Obviously, I cannot merge a version with a commit message called "please, only review - don't merge.".

File: cql3/query_processor.cc
Comment (nyh) [reply cql3/query_processor.cc:354]: @vladzcloudius I need to make something clear about my review - and all my reviews in general.
When I say that something is not clear, it doesn't always mean I want an explanation now. It often means that I thought that the code or error message won't be clear to future users or reader as well - not just to me. In this case I noted that a message like "this is a red flag" sounds scary but doesn't help understand what's wrong. If the problem is that the user is probably preparing different statements and then using them only once - why not say exactly this? E.g.,

"Counts a number of evictions of prepared statements from the prepared statements cache after they have been used only once. An increasing counter suggests the user may be preparing a different statement for each request instead of reusing the same prepared statement with parameters".

By the way, the other case where the user is preparing exactly the same statement over and over with exactly the same string, do we have a counter for that too? Isn't this a problem too?

**Comment by vladzcloudius** [comment]
> @vladzcloudius please clean up whatever you want to clean up. Obviously, I cannot merge a version with a commit message called "please, only review - don't merge.".

Sure. I'll create a clean PR and tag you there. Thanks for a review, @nyh

File: cql3/query_processor.cc
Comment (vladzcloudius) [reply cql3/query_processor.cc:354]: > By the way, the other case where the user is preparing exactly the same statement over and over with exactly the same string, do we have a counter for that too? Isn't this a problem too?

@nyh It's not really an issue in a general case because it could be happening simply because the user has a worker that uses the same (good) parametrized query but the worker is short lived, so instances of that worker are being created over and over again and then they finish and exit. 

Another case similar to the above is when multiple instances of that worker are running concurrently and the first thing a single worker does is calling a PREPARE.

We already have this patch very optimized. In that case no CQL parsing is expected for the second time PREPARE is called.

**Pull request closed without merging**
